security/pkg/pki/ra/k8s_ra.go

```diff
@@ -94,10 +98,73 @@ func (r *KubernetesRA) SignWithCertChain(csrPEM []byte, certOpts ca.CertOpts) ([
 	if len(chainPem) > 0 {
 		cert = append(cert, chainPem...)
 	}
-	return cert, nil
+	respCertChain := []string{string(cert)}
+	var rootCert, rootCertFromMeshConfig, rootCertFromCertChain []byte
+	certSigner := r.certSignerDomain + "/" + certOpts.CertSigner
+	if len(r.GetCAKeyCertBundle().GetRootCertPem()) == 0 {
+		rootCertFromCertChain, err = util.FindRootCertFromCertificateChainBytes(cert)
+		if err != nil {
+			return nil, fmt.Errorf("failed to find root cert from signed cert-chain (%v)", err.Error())
+		}
+		rootCertFromMeshConfig, err = r.GetRootCertFromMeshConfig(certSigner)
+		if err != nil {
+			return nil, fmt.Errorf("failed to find root cert from mesh config (%v)", err.Error())
+		}
+		if rootCertFromCertChain != nil && rootCertFromMeshConfig != nil {
+			if string(rootCertFromCertChain) != string(rootCertFromMeshConfig) {
+				return nil, fmt.Errorf("root cert from signed cert-chain" +
+					" is conflicting with the one specified in mesh config")
+			}
+		}
+		if rootCertFromMeshConfig != nil {
+			rootCert = rootCertFromMeshConfig
+		}
+
+		if rootCertFromCertChain != nil {
+			rootCert = rootCertFromCertChain
+		}
+
+		if verifyErr := util.VerifyCertificate(nil, cert, rootCert, nil); verifyErr != nil {
+			return nil, fmt.Errorf("root cert from signed cert-chain is invalid %v ", verifyErr)
+		}
+		respCertChain = append(respCertChain, string(rootCert))
+	}
+	return respCertChain, nil
 }
 
 // GetCAKeyCertBundle returns the KeyCertBundle for the CA.
 func (r *KubernetesRA) GetCAKeyCertBundle() *util.KeyCertBundle {
 	return r.keyCertBundle
 }
+
+func (r *KubernetesRA) GetCACertificatesFromMeshConfig() []*meshconfig.MeshConfig_CertificateData {
+	return r.caCertificatesFromMeshConfig
```

Comment:
Two points: I am not sure it is safe to hold a reference to meshconfig.MeshConfig_CertificateData here. Would be safer to copy over the certificates into an array here IMO. Secondly, the meshHandler code and GetCACertificatesFromMeshConfig are called from different goroutines. We need to have a mutex here IMO.